src/Helpers/TenantHelper.php

```diff
@@ -0,0 +1,54 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Slides\Saml2\Helpers;
+
+use Slides\Saml2\Models\Tenant;
+use Illuminate\Support\Facades\URL;
+
+/**
+ * TenantHelper wraps a Tenant and adds business logic.
+ * 
+ * Done this way to the business logic can be unit tested, on mock Tenant,
+ * since Eloquent models themselves are virtually impossible to unit test.
+ *
+ * @package App\Helpers
+ */
+class TenantHelper
+{
+    private Tenant $tenant;
+
+    final public function __construct(Tenant $tenant)
+    {
+        $this->tenant = $tenant;
+    }
+
+    /**
+     * Factory method for fluent calls,
+     * e.g. TenantHelper::with($tenant)->blah_blah
+     *
+     * @param Tenant $tenant
+     * @return TenantHelper
+     */
+    public static function with(Tenant $tenant): self
+    {
+        return new static($tenant);
+    }
+
+    /**
+     * Use this to determine the SP Entity ID for a tenant.
+     *
+     * 1. Returns default SP entity ID (metadata URL) *UNLESS* sp_entity_id_override is filled in (non-empty string),
+     * 2. ... in which case it will return that sp_entity_id_override value
+     * 
+     * Q: Why not just make this Tenant::getSpEntityIdAttribute?
+     * A: Because Eloquent models hate unit testing
+     *
+     * @return string SP Entity ID
+     */
+    public function getSpEntityId(): string
+    {
+        return $this->tenant->sp_entity_id_override ?: URL::route('saml.metadata', ['uuid' => $this->tenant->uuid]);
```

Comment:
The way I understand this the override will be the full URL. That means if anything changes regarding URLs and routing this would be out of date.

Have considered only keeping the domain URI part in the db.

I think it would be something like this.

```suggestion
        $route = URL::route('saml.metadata', ['uuid' => $this->tenant->uuid]);

        if ($this->tenant->sp_entity_id_override) {
            $route->domain($this->tenant->sp_entity_id_override);
        }

        return $route; 
```

Reply:
> Have considered only keeping the domain URI part in the db.

Yes I considered that, but I figured a more "general purpose SP Entity ID override" is better for possible library reused, and Entity IDs do *not* need to be metadata URL served by the app.

It need to be a URI (usually a URL) but technically
> The domain in the entity ID [does not] need [to] match those in the endpoint locations in metadata
> https://spaces.at.internet2.edu/display/federation/saml-metadata-entityid

Also... from that same article
> Make every effort to choose an entity ID that will persist indefinitely

is making me think, is there *any* way we can do that?

(I think I did find the answer though and it's no, because the library matches incoming request by 
GUID and finds the *first* tenant that matches that guid........
⋮
⋮
⋮
⋮
⋮
⋮
WAIT.

That's the GUID of the ACS.

Because of the hardcoded above, the SP Entity ID guid does NOT have to be real, does not have to match the ACS guid.....
It MIGHT be possible for the client to ***not have to change the SP Entity ID***, just change the ACS (to different domain AND different guid)......!!

Wow, that got me thinking!!

Thanks!

But in summary "yeah I think it's actually better that it can be hardcoded anything string!"